go-controller/pkg/util/multi_network.go

```diff
@@ -761,6 +761,61 @@ func ValidateNetConf(nadName string, netconf *ovncnitypes.NetConf) error {
 		return fmt.Errorf("the subnet attribute must be defined for layer2 primary user defined networks")
 	}
 
+	if netconf.Topology != types.LocalnetTopology && netconf.Name != types.DefaultNetworkName {
+		if err := subnetOverlapCheck(netconf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// subnetOverlapCheck validates whether POD and join subnet mentioned in a net-attach-def with
+// topology "layer2" and "layer3" does not overlap with ClusterSubnets, ServiceCIDRs, join subnet,
+// and masquerade subnet. It also considers excluded subnets mentioned in a net-attach-def.
+func subnetOverlapCheck(netconf *ovncnitypes.NetConf) error {
+	allSubnets := config.NewConfigSubnets()
+	for _, subnet := range config.Default.ClusterSubnets {
+		allSubnets.Append(config.ConfigSubnetCluster, subnet.CIDR)
+	}
+	for _, subnet := range config.Kubernetes.ServiceCIDRs {
+		allSubnets.Append(config.ConfigSubnetService, subnet)
+	}
+	_, v4JoinCIDR, _ := net.ParseCIDR(config.Gateway.V4JoinSubnet)
+	_, v6JoinCIDR, _ := net.ParseCIDR(config.Gateway.V6JoinSubnet)
+
+	allSubnets.Append(config.ConfigSubnetJoin, v4JoinCIDR)
+	allSubnets.Append(config.ConfigSubnetJoin, v6JoinCIDR)
+
+	_, v4MasqueradeCIDR, _ := net.ParseCIDR(config.Gateway.V4MasqueradeSubnet)
+	_, v6MasqueradeCIDR, _ := net.ParseCIDR(config.Gateway.V6MasqueradeSubnet)
+
+	allSubnets.Append(config.ConfigSubnetMasquerade, v4MasqueradeCIDR)
+	allSubnets.Append(config.ConfigSubnetMasquerade, v6MasqueradeCIDR)
+
+	ni, err := NewNetInfo(netconf)
+	if err != nil {
+		return err
+	}
+	for _, subnet := range ni.Subnets() {
+		allSubnets.Append(config.UserDefinedSubnets, subnet.CIDR)
+	}
+
+	for _, subnet := range ni.JoinSubnets() {
+		allSubnets.Append(config.UserDefinedJoinSubnet, subnet)
+	}
+	if ni.ExcludeSubnets() != nil {
```

Comment:
what is the reason for having this block ? aren't we just good enough by not adding `ni.ExcludeSubnets()` into allSubnets before doing overlap check ?

Reply:
https://github.com/openshift/enhancements/pull/1623/files#diff-76f8949dbcefb73686f569b18267b152f0f1ef3d71ce9888046aa93f97566005R319 makes me think that we need to run overlap check for join subnet mentioned at a UDN.

Reply:
ok, if this parameter is meant to exclude matching subnets from overlap check, then i'm fine with it.

Reply:
Yes, that is my intention here. Thanks!

Reply:
@maiqueb / @qinqon when validating exclude subnets.. we don't want to check for overlap with used subnets? I thought we should..

Reply:
ah nvm this exclude subnets is a field where the user provides the values that should be excluded from IPAM and usually its actually contained within the main subnet right?

Reply:
Yes, the overlap is expected. We want to have a way for users to express they want to ensure some IPs are never addressable within their subnets.

Reply:
> ah nvm this exclude subnets is a field where the user provides the values that should be excluded from IPAM and usually its actually contained within the main subnet right? so the overlap is expected?

Yes, overlap with either POD subnet or join subnet in a NAD is expected.